bin/lint-markdown-api-history.ts

```diff
@@ -0,0 +1,273 @@
+#!/usr/bin/env node
+
+import { readFile, access, constants } from 'node:fs/promises';
+import { resolve } from 'node:path';
+import * as minimist from 'minimist';
+
+import { DocsWorkspace } from '../lib/markdown';
+import { URI } from 'vscode-uri';
+import { parse as parseYaml } from 'yaml';
+
+import type { HTML } from 'mdast';
+import type { Node, Literal } from 'unist';
+import type { visit as VisitFunction } from 'unist-util-visit';
+import type { fromMarkdown as FromMarkdownFunction } from 'mdast-util-from-markdown';
+import { dynamicImport } from '../lib/helpers';
+import Ajv, { JSONSchemaType, ValidateFunction } from 'ajv';
+
+const apiHistoryRegex: RegExp = /<!--\r?\n(```YAML history\r?\n([\s\S]*?)\r?\n```)\r?\n-->/g;
```

Comment:
I don't think we should use a regex for this purpose, as it's not very robust and doesn't have user friendly errors when it doesn't match (what you call format errors in this PR). Things like trailing whitespace, extra newlines after the opening `<!--`, etc, will fail using this regex, and it's not straightforward to separate out the failure cases.

Here's a quick possible alternative implementation which is more robust to those situations. It's surely slower than the simple regex (it has two more parse steps, `fromHtml` and another `fromMarkdown`), but we'd be trading some performance for robustness.

It uses `hast-util-from-html` to robustly parse the HTML comment so that just the content can be retrieve (unfortunately `mdast-util-from-markdown` only gives the raw HTML content) - alternatively this could be replaced with a simple regex to extract the content inside `<!-- -->`, but let's start with the most robust solution and back off to a simpler solution if we decide it's necessary for performance reasons.

Then the code block is parsed using `fromMarkdown`.

This should let us add some human friendly output for format error situations, for example (not suggesting all of these be implemented in this PR, merely illustrating linting checks we could add):
* Error if there's no extraneous whitespace around `history` (the meta string for the code block)
* Error if the lang is not uppercase `YAML` (there's a similar linting check in `e/e` to enforce consistent lang identifiers for code blocks)
* Error if the parsed Markdown inside the HTML comment finds has extraneous children (i.e. it should only have the code block)
* Error regarding trailing whitespace, extra newlines, etc.

```diff
diff --git a/bin/lint-markdown-api-history.ts b/bin/lint-markdown-api-history.ts
index f1a1b28..cc0c117 100644
--- a/bin/lint-markdown-api-history.ts
+++ b/bin/lint-markdown-api-history.ts
@@ -11,12 +11,11 @@ import { parse as parseYaml } from 'yaml';
 import type { HTML } from 'mdast';
 import type { Node, Literal } from 'unist';
 import type { visit as VisitFunction } from 'unist-util-visit';
+import type { fromHtml as FromHtmlFunction } from 'hast-util-from-html';
 import type { fromMarkdown as FromMarkdownFunction } from 'mdast-util-from-markdown';
 import { dynamicImport } from '../lib/helpers';
 import Ajv, { JSONSchemaType, ValidateFunction } from 'ajv';
 
-const apiHistoryRegex: RegExp = /<!--\r?\n(```YAML history\r?\n([\s\S]*?)\r?\n```)\r?\n-->/g;
-
 interface ChangeSchema {
   'pr-url': string;
   'breaking-changes-header'?: string;
@@ -80,6 +79,13 @@ async function main(
     ignoreGlobs = [],
   }: Options,
 ) {
+  const { fromHtml } = (await dynamicImport('hast-util-from-html')) as {
+    fromHtml: typeof FromHtmlFunction;
+  };
+  const { fromMarkdown } = (await dynamicImport('mdast-util-from-markdown')) as {
+    fromMarkdown: typeof FromMarkdownFunction;
+  };
+
   let documentCounter = 0;
   let historyBlockCounter = 0;
   let errorCounter = 0;
@@ -111,10 +117,20 @@ async function main(
     for (const possibleHistoryBlock of possibleHistoryBlocks) {
       historyBlockCounter++;
 
-      const regexMatchIterator = possibleHistoryBlock.value.matchAll(apiHistoryRegex);
-      const regexMatches = Array.from(regexMatchIterator);
-
-      if (regexMatches.length !== 1 || regexMatches[0].length !== 3) {
+      const {
+        children: [htmlComment],
+      } = fromHtml(possibleHistoryBlock.value);
+      if (htmlComment.type !== 'comment') {
+        continue;
+      }
+      const {
+        children: [codeBlock],
+      } = fromMarkdown(htmlComment.value);
+      if (
+        codeBlock.type !== 'code' ||
+        codeBlock.lang?.toLowerCase() !== 'yaml' ||
+        codeBlock.meta?.trim() !== 'history'
+      ) {
         console.error(
           `Error parsing ${filepath}\n
           Couldn't extract matches from possible history block, did you use the correct format?:\n
@@ -124,12 +140,10 @@ async function main(
         continue;
       }
 
-      const historyYaml = regexMatches[0][2];
-
       let unsafeHistory = null;
 
       try {
-        unsafeHistory = parseYaml(historyYaml);
+        unsafeHistory = parseYaml(codeBlock.value);
       } catch (error) {
         console.error(
           `Error parsing\n
```

Reply:
Sounds good, applied you patch in [`9038174` (#73)](https://github.com/electron/lint-roller/pull/73/commits/9038174ebe8b7e3eeb17b619881a208aa25b8a83)